torch/distributed/tensor/parallel/style.py

```diff
@@ -1,652 +1,357 @@
 
-        super().__init__(
-            _prepare_input=prepare_input_fn,
-            _prepare_output=prepare_output_fn,
-            input_layouts=input_layouts,
-            output_layouts=output_layouts,
-            use_local_output=use_local_output,
+    def _prepare_input_fn(self, inputs, device_mesh):
+        dtensor = DTensor.from_local(inputs[0], device_mesh, self.input_layouts)
+        if self.input_layouts != self.desired_input_layouts:
+            dtensor = dtensor.redistribute(placements=self.desired_input_layouts)
+        return dtensor
+
+    def _partition_fn(self, name, module, device_mesh):
+        assert isinstance(module, nn.Linear), "Only support nn.Linear"
+        # Rowwise shard weight to Shard(1), bias to Replicate(), weight be Shard(1)
+        # means Rowwise as Linear is input * weight^T + bias, where
+        # weight would become Shard(0)
+        module.register_parameter("weight", nn.Parameter(
+            distribute_tensor(module.weight, device_mesh, [Shard(1)])
+        ))
+        if module.bias is not None:
+            module.register_parameter("bias", nn.Parameter(
+                distribute_tensor(module.bias, device_mesh, [Replicate()])
+            ))
+
+    def _prepare_output_fn(self, outputs, device_mesh):
+        if outputs.placements != self.desired_output_layouts:
+            outputs = outputs.redistribute(placements=self.desired_output_layouts)
+        # back to local tensor if use_local_output is True
+        return outputs.to_local() if self.use_local_output else outputs
+
+    def apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        return distribute_module(
+            module,
+            device_mesh,
+            self._partition_fn,
+            self._prepare_input_fn,
+            self._prepare_output_fn
         )
 
-    @staticmethod
-    def _prepare_input_fn(input_layouts, output_layouts, inputs, device_mesh=None):
-        return (
-            _redistribute_per_both_layouts(
-                inputs[0], input_layouts, output_layouts, device_mesh
-            ),
-        )
 
 class PrepareModuleInput(ParallelStyle):
     """
-    Annotate Tensor inputs with layouts for conversion to DTensor, redistributing based on specified output layouts.
-
-    :class:`PrepareModuleInput` enables users to annotate :class:`torch.Tensor` or :class:`DTensor`
-    inputs with ``input_layouts`` and ``output_layouts`` so that each input can be converted to
-    :class:`DTensor` based on the annotation. Specifically, a DTensor will be created
-    from the input Tensor based on ``input_layouts`` and then redistributed to another
-    DTensor based on ``output_layouts``.
-
-    When the input is not a :class:`torch.Tensor` or :class:`DTensor`, if no layout is
-    specified, it will be a no-op. Otherwise, it will throw an error.
+    Annotate the input tensors of nn.Module to DTensors according to input_layouts, and perform layout redistribution
```

Comment:
the verb here isn't quite clear.  Annotate sounds like it just marks something.  But "to DTensors" sounds like it made a change (conversion).  Would it be more correct to say "Convert the input tensors of nn.Module to DTensors"?

or maybe this is more correct? 

"Configure the nn.Module's to convert its inputs to DTensor at runtime following the parallelize plan"